Azure/sonic-mgmt
# [dualtor] Add `nic_simulator`

### Description of PR


Summary:
Fixes # (issue)

### Type of change





### Back port request

### Approach
#### What is the motivation for this PR?
Add `nic_simulator` to act as gRPC server to respond to requests from DUT operates in dualtor `active-active` mode.

Signed-off-by: Longxiang Lyu <lolv@microsoft.com>

#### How did you do it?
Add script `nic_simulator.py` to start per-port gRPC server for a dualtor `active-active` setup testbed. It should run in the server NiCs namespace, and it basically does the following few things after it starts:
1. scans for all the ports in the namespace that follows the server NiC name pattern(`eth%d`) and finds its corresponding OVS bridge, the bridge will be verified to have the following topology:
```
                                   +--------------+
                   PTF (host_if) --+              +----- upper_if
                                   |  OVS bridge  |
    simulator netns (server_nic) --+              +----- lower_if
                                   +--------------+
```
2. starts to set up the flows for each OVS bridge collected in step1, the initial forwarding state is `(active, active)`.
    * the flows added:
        * `FLOW I`: for packets from `upper_if`, it will be duplicated to `host_if` and `server_nic`.
        * `FLOW II`: for packets from `lower_if`, it will be duplicated to `host_if` and `server_nic`.
        * `FLOW III`: for packets from `server_nic`, it will be duplicated to `upper_if` and `lower_if`.
        * `FLOW IV`: for packets from `host_if`, it will be directed to either `upper_if` or `lower_if`.

3. starts the gRPC services(one port, one server) to query/set the forwarding state of the bridge.
    * if upper ToR forwarding state is set to `standby`, `nic_simulator` will modify `FLOW IV` to let packets from `host_if` only be directed to lower ToR(`active`)

**TODO**
1. Add RPC calls to cover testing scenarios like link drop and link down.
2. Add client authentication.
3. Implement `RPC` call `QueryOperationPortState`
4. add logging interceptors to the RPC calls.

#### How did you verify/test it?
1. run the `nic_simulator.py`
2. run the `nic_simulator_client.py` to verify `nic_simulator` operations.

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation
